src/math/trig.c

```diff
@@ -9,568 +9,115 @@
+	/* 1686629713, deg = 90.000000  */
+	const int32_t CORDICSINE_PIOVERTWO_Q30FL  = Q_CONVERT_FLOAT(M_PI / 2, 30);
+	/*-1073741824, deg = -57.2958*/
+	/* Minor-Bug,-1073741824 = round(-1*2^30) but QCF is -1073741823*/
+	const int32_t CORDICSINE_N57DEG_Q30FL = Q_CONVERT_FLOAT(-1, 30);
 
-	i1 = idx & (2 * SINE_NQUART - 1);
-	if (i1 > SINE_NQUART)
-		i1 = 2 * SINE_NQUART - i1;
+	/*Addition or subtraction by a multiple of pi/2 is done in the data type
+	 *of the input. When the fraction length is 29, then the quantization error
+	 *introduced by the addition or subtraction of pi/2 is done with 29 bits of
+	 *precision.Input range of cordicsin must be in the range [-2*pi, 2*pi),
+	 *a signed type with fractionLength = wordLength-4 will fit this range
+	 *without overflow.Increase of fractionLength makes the addition or
+	 *subtraction of a multiple of pi/2 more precise
+	 */
+	if (th_rad_fxp > CORDICSINE_PIOVERTWO_Q28FL) {
+		if ((th_rad_fxp - CORDICSINE_PIOVERTWO_Q29FL) <= CORDICSINE_PIOVERTWO_Q28FL) {
+			th_rad_fxp -= CORDICSINE_PIOVERTWO_Q29FL;
+			sign  = -1;
+		} else {
+			th_rad_fxp -= CORDICSINE_PIOVERTWO_Q30FL;
+			sign  = 1;
+		}
+	} else if (th_rad_fxp < -CORDICSINE_PIOVERTWO_Q28FL) {
+		if ((th_rad_fxp + CORDICSINE_PIOVERTWO_Q29FL) >= -CORDICSINE_PIOVERTWO_Q28FL) {
+			th_rad_fxp += CORDICSINE_PIOVERTWO_Q29FL;
+			sign  = -1;
+		} else {
+			th_rad_fxp += CORDICSINE_PIOVERTWO_Q30FL;
+			sign  = 1;
+		}
+	} else {
+		sign  = 1;
+	}
+	th_rad_fxp <<= 2;
+	b_yn = 0;
+	xn = CORDICSINE_LUT_Q29FL;
+	xtmp = CORDICSINE_LUT_Q29FL;
+	ytmp = 0;
+	for (b_idx = 0; b_idx < 31; b_idx++) {
+		if (th_rad_fxp < 0) {
+			th_rad_fxp += cordicsine_lookup[b_idx];
+			xn += ytmp;
+			b_yn -= xtmp;
+		} else {
+			th_rad_fxp -= cordicsine_lookup[b_idx];
+			xn -= ytmp;
+			b_yn += xtmp;
+		}
+		xtmp = xn >> (b_idx + 1);
+		ytmp = b_yn >> (b_idx + 1);
+	}
 
-	if (idx > 2 * SINE_NQUART)
-		s = -sine_table[i1];
+	if (sign == -1)
+		th_rad_fxp = sign * b_yn;
 	else
-		s = sine_table[i1];
-
-	return s;
-}
-
-/* Compute fixed point sine with table lookup and interpolation */
-int32_t sin_fixed(int32_t w)
-{
-	int32_t frac;
-	int32_t s0;
-	int32_t s1;
-	int32_t delta;
-	int64_t idx;
-	int64_t sine;
-	int64_t idx_tmp;
+		th_rad_fxp = sign * b_yn;
+	/*To support Q1.31 format, look for +/-57deg*/
+	if (th_rad_fxp > CORDICSINE_P57DEG_Q30FL)
+		th_rad_fxp = INT32_MAX;
+	else if (th_rad_fxp <= CORDICSINE_N57DEG_Q30FL)
+		th_rad_fxp = INT32_MIN;
+	else
+		th_rad_fxp <<= 1;
 
-	/* Q4.28 x Q12.20 -> Q16.48 */
-	idx_tmp = (int64_t)w * SINE_C_Q20;
-	idx = (idx_tmp >> 48); /* Shift to Q0 */
-	idx_tmp = idx_tmp >> 17; /* Shift to Q16.31 */
-	idx_tmp = idx_tmp - (idx << 31); /* Get fraction */
-	frac = (int32_t)idx_tmp; /* Q1.31 */
-	s0 = sine_lookup(idx); /* Q1.31 */
-	s1 = sine_lookup(idx + 1); /* Q1.31 */
-	delta = s1 - s0; /* Q1.31 */
-	/* All Q1.31 */
-	sine = s0 + q_mults_32x32(frac, delta, Q_SHIFT_BITS_64(31, 31, 31));
+	cdc_sin_th = th_rad_fxp;
+	return cdc_sin_th;
```

Comment:
why not just return th_rad_fxp?

Reply:
Yep, this can be done.
Is done to keep input and return value name different

Reply:
i dont think it matters, adding extra values looks like a bug.

Reply:
Fine , then I return th_rad_fxp instead of cdc_sin_th